### PR TITLE
fix: make TestMempoolProgressInHigherRound less flaky

### PR DESCRIPTION
**Pull request by faddat**
This PR should make TestMempoolProgressInHigherRound less flaky and a teensy big more rigorous. 

The critical change is the order of the steps in the test itself, which does seem to make the test more reliable.

If the order should not have changed then we should probably just close this PR and try to find another way to make the test more reliable.

closes: #1932 

found in: #1954 

---

#### PR checklist

- [x] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: internal/consensus/mempool_test.go [hunk 1]
```diff
@@ -70,41 +70,58 @@ func TestMempoolProgressAfterCreateEmptyBlocksInterval(t *testing.T) {
...
+
+	// Subscribe to events
 	newBlockCh := subscribe(cs.eventBus, types.EventQueryNewBlock)
 	newRoundCh := subscribe(cs.eventBus, types.EventQueryNewRound)
 	timeoutCh := subscribe(cs.eventBus, types.EventQueryTimeoutPropose)
+
+	// Set proposal function
 	cs.setProposal = func(proposal *types.Proposal) error {
 		if cs.Height == 2 && cs.Round == 0 {
-			// dont set the proposal in round 0 so we timeout and
```
Comment (melekes) [inline internal/consensus/mempool_test.go:86]: why this was removed?

**Review by melekes — COMMENTED** [review]
Thanks @faddat ❤️ Unsure how changing/adding comments fixes the test's flackiness though. Could you please explain?

File: internal/consensus/mempool_test.go (hunk 1 above)
Comment (faddat) [reply internal/consensus/mempool_test.go:86]: no good reason and hapy to restore it

Comment (faddat) [reply internal/consensus/mempool_test.go:86]: it's back

**Comment by faddat** [comment]
wrangling.... will mark this as draft.

**Comment by faddat** [comment]
That's right draft is a good way for this one.

File: privval/socket_listeners_test.go
Comment (cason) [inline privval/socket_listeners_test.go:1]: Are these changes somehow related with the mempool test?

File: internal/consensus/mempool_test.go [hunk 2]
```diff
@@ -90,21 +100,30 @@ func TestMempoolProgressInHigherRound(t *testing.T) {
 		}
 		return cs.defaultSetProposal(proposal)
 	}
+
+	// Start test round
+	height, round := cs.Height, cs.Round
```
[the remark this answers is not in the document]
Comment (cason) [reply internal/consensus/mempool_test.go:105]: Don't see why. Height is always 1, round is always 0.

File: internal/consensus/mempool_test.go [hunk 3]
```diff
@@ -90,21 +100,30 @@ func TestMempoolProgressInHigherRound(t *testing.T) {
...
 	startTestRound(cs, height, round)
 
-	ensureNewRound(newRoundCh, height, round) // first round at first height
-	ensureNewEventOnChannel(newBlockCh)       // first block gets committed
+	// Ensure new round at first height and block commit
+	ensureNewRound(newRoundCh, height, round)
```
Comment (cason) [inline internal/consensus/mempool_test.go:109]: The problem here is that the new round event takes more than 200ms?

Comment (faddat) [reply internal/consensus/mempool_test.go:109]: These are both good questions that I had missed. I will review this tomorrow and thank you for your review

**Comment by faddat** [comment]
@cason I think we should defer further review here until https://github.com/cometbft/cometbft/pull/2021

is merged.

**Comment by faddat** [comment]
@cason please see #1923 

I just added some commentary about another time-related test issue and I'm not sure how to handle these safely in all cases.
